src/crypto/ocb.js

```diff
@@ -0,0 +1,332 @@
+// OpenPGP.js - An OpenPGP implementation in javascript
+// Copyright (C) 2018 ProtonTech AG
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 3.0 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+/**
+ * @fileoverview This module implements AES-OCB en/decryption.
+ * @requires crypto/cipher
+ * @requires util
+ * @module crypto/ocb
+ */
+
+import ciphers from './cipher';
+import util from '../util';
+
+
+const blockLength = 16;
+const ivLength = 15;
+
+// https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-04#section-5.16.2:
+// While OCB [RFC7253] allows the authentication tag length to be of any
+// number up to 128 bits long, this document requires a fixed
+// authentication tag length of 128 bits (16 octets) for simplicity.
+const tagLength = 16;
+
+
+const { shiftRight } = util;
+
+
+function ntz(n) {
+  let ntz = 0;
+  for(let i = 1; (n & i) === 0; i <<= 1) {
+    ntz++;
+  }
+  return ntz;
+}
+
+function xorMut(S, T) {
+  for (let i = 0; i < S.length; i++) {
+    S[i] ^= T[i];
+  }
+  return S;
+}
+
+function xor(S, T) {
+  return xorMut(S.slice(), T);
+}
+
+/**
+ * If S[1] == 0, then double(S) == (S[2..128] || 0);
+ * otherwise, double(S) == (S[2..128] || 0) xor
+ * (zeros(120) || 10000111).
+ *
+ * The only explicitly conditional OCB operation that depends on private
+ * data is double(), which means that using constant-time blockcipher
+ * and double() implementations eliminates most (if not all) sources of
+ * timing attacks on OCB.
+ *
+ * @param {Uint8Array} data (16 bytes)
+ */
+function double(data) {
```

Comment:
Can't this function be shared between here and the cmac code? Probably better than maintaining two nearly identical pieces of code